experimental/gittuf/hook.go

```diff
@@ -22,8 +33,148 @@ func (e *ErrHookExists) Error() string {
+// user defined by principalID. Upon successful completion of all hooks for the
+// stage for the user, the map of hook names to exit codes is returned.
+// TODO: Add attestations workflow
+func (r *Repository) InvokeHooksForStage(ctx context.Context, stage tuf.HookStage, signer sslibdsse.Signer, opts ...hookopts.Option) (map[string]int, error) {
+	options := &hookopts.Options{}
+	for _, fn := range opts {
+		fn(options)
+	}
+
+	// TODO: Use signerverifier API to look at Git config if no signer specified
+	if signer == nil {
+		return nil, sslibdsse.ErrNoSigners
+	}
+
+	keyID, err := signer.KeyID()
+	if err != nil {
+		return nil, err
+	}
+
+	slog.Debug("Loading current policy...")
+	state, err := policy.LoadCurrentState(ctx, r.r, policy.PolicyRef)
+	if err != nil {
+		return nil, err
+	}
+
+	rootMetadata, err := state.GetRootMetadata(false)
+	if err != nil {
+		return nil, err
+	}
+
+	allHooks, err := rootMetadata.GetHooks(stage)
+	if err != nil {
+		return nil, err
+	}
+
+	var selectedHooks []tuf.Hook
+	var selectedPrincipal tuf.Principal
+
+	// Read the principals from targetsMetadata and attempt to find a match for
+	// the specified principal to determine which hooks to run.
+	for _, principal := range state.GetAllPrincipals() {
+		for _, key := range principal.Keys() {
+			if key.KeyID == keyID {
+				selectedPrincipal = principal
+				break
+			}
+		}
+	}
+
+	// Couldn't match the key up to a principal, abort
+	if selectedPrincipal == nil {
+		return nil, tuf.ErrPrincipalNotFound
+	}
+
+	// Now, read all hooks for the specified stage and find which ones we need
+	// to run
+	for _, hook := range allHooks {
+		principalIDs := hook.GetPrincipalIDs()
+		if principalIDs.Has(selectedPrincipal.ID()) {
+			selectedHooks = append(selectedHooks, hook)
+		}
+	}
+
+	if len(selectedHooks) == 0 {
+		return nil, ErrNoHooksFoundForPrincipal
+	}
+
+	// Determine what parameters must be supplied based on the hook stage
+	var luaParameters lua.LTable
+
+	// At the moment, the only stage that we support that requires parameters is
+	// the pre-push stage.
+	if stage == tuf.HookStagePrePush {
+		// https://git-scm.com/docs/githooks#_pre_push
+		// For pre-push hooks, we supply two things:
+		// 1. The remote name and destination, e.g.
+		// origin git@github.com:gittuf/gittuf
+		// 2. The local and remote refs/object IDs in the form:
+		// <local ref> <local hash> <remote ref> <remote hash>
+
+		luaParameters.RawSet(lua.LString("remoteName"), lua.LString(options.RemoteName))
+		luaParameters.RawSet(lua.LString("remoteURL"), lua.LString(options.RemoteURL))
+
+		remoteObjects := make(map[string][]gitinterface.Hash, len(options.RefSpecs))
+
+		for _, refSpec := range options.RefSpecs {
+			splitRefSpec := strings.Split(refSpec, ":")
+			localRef, remoteRef := splitRefSpec[0], splitRefSpec[1]
+
+			var remoteHash gitinterface.Hash
+
+			err = r.r.Fetch(options.RemoteName, []string{remoteRef}, true)
```

Comment:
isn't this going to try and fetch into a branch we're pushing? don't we want this fetch to be into remote tracker for the branch?